File: 02_time_series_plots.py
#!/usr/bin/env python3
"""
==============================================================================
Stage 2: Time Series Plots
==============================================================================
Plot 2a: Netanyahu's bloc (Likud + Religious Zionism) — single trend line
         with election vertical line
Plot 2b: Rightwards vs Israel Our Home — election vertical line
Outputs: output/plot2a_radicalized_trend.png
         output/plot2b_change_coalition_trend.png
==============================================================================
"""

import pandas as pd
import numpy as np
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import matplotlib.ticker as mtick
import seaborn as sns

sns.set_style("whitegrid")
plt.rcParams["figure.dpi"] = 300
plt.rcParams["savefig.dpi"] = 300

WINDOW   = 4          # weeks rolling average
ELEC_WK  = 0          # election is week 0 by definition

print("=" * 70)
print("STAGE 2: TIME SERIES PLOTS")
print("=" * 70)

data = pd.read_pickle("output/analysis_data.pkl")
print(f"\nLoaded {len(data):,} rows")

# ── helper ───────────────────────────────────────────────────────────────────
def weekly_prop(df_subset):
    """Return weekly proportion of populist tweets."""
    return (df_subset.groupby("week_from_election")
                     .agg(pop_sum=("pop", "sum"), total=("pop", "count"))
                     .assign(prop=lambda x: x.pop_sum / x.total)
                     .reset_index())

def smooth(series, w=WINDOW):
    return series.rolling(window=w, center=True, min_periods=1).mean()

# ──────────────────────────────────────────────────────────────────────────────
# PLOT 2a — Netanyahu's bloc — single trend line, composition changes at cutoff
#   Pre-election : named PRR legislators (prr_leg_pre screen_names) + all Likud
#   Post-election: all Religious Zionism (party) + all Likud
# ──────────────────────────────────────────────────────────────────────────────
print("\nPlot 2a: Netanyahu's bloc trend (single line, correct pre/post composition)...")

bloc_mask = (
    ((data["post_election"] == 0) & (data["prr_leg_pre"] | (data["party"] == "Likud"))) |
    ((data["post_election"] == 1) & (data["radicalized_group"]))
)
rad_wk = weekly_prop(data[bloc_mask]).sort_values("week_from_election")

fig, ax = plt.subplots(figsize=(12, 6))

# scatter dots (faint)
ax.scatter(rad_wk["week_from_election"], rad_wk["prop"],
           alpha=0.18, s=14, color="#c0392b")

# single smoothed line
ax.plot(rad_wk["week_from_election"], smooth(rad_wk["prop"]),
        linewidth=2.5, color="#c0392b",
        label="Netanyahu's Bloc (Likud + Religious Zionism)")

# election line
ax.axvline(x=0, color="#2c3e50", linestyle="--", linewidth=2,
           label="Election — March 23, 2021")

ax.yaxis.set_major_formatter(mtick.PercentFormatter(xmax=1, decimals=1))
ax.set_xlabel("Weeks from Election (March 23, 2021)", fontweight="bold")
ax.set_ylabel("Proportion of Populist Tweets", fontweight="bold")
ax.set_title("Populism in Netanyahu's Bloc Before and After 2021 Election",
             fontweight="bold", pad=14)
ax.legend(loc="upper left", frameon=True, fancybox=True, shadow=True, fontsize=10)
ax.grid(True, alpha=0.3)
plt.tight_layout()
plt.savefig("output/plot2a_radicalized_trend.png", bbox_inches="tight")
plt.close()
print("✓ output/plot2a_radicalized_trend.png")

# ──────────────────────────────────────────────────────────────────────────────
# PLOT 2b — Rightwards vs Israel Our Home
# ──────────────────────────────────────────────────────────────────────────────
print("\nPlot 2b: Rightwards vs Israel Our Home trend...")

right_wk = weekly_prop(data[data["party"] == "Rightwards"]).sort_values("week_from_election")
ioh_wk   = weekly_prop(data[data["party"] == "Israel Our Home"]).sort_values("week_from_election")

fig, ax = plt.subplots(figsize=(12, 6))

ax.scatter(right_wk["week_from_election"], right_wk["prop"],
           alpha=0.18, s=14, color="#2980b9")
ax.scatter(ioh_wk["week_from_election"], ioh_wk["prop"],
           alpha=0.18, s=14, color="#27ae60")

ax.plot(right_wk["week_from_election"], smooth(right_wk["prop"]),
        linewidth=2.5, color="#2980b9", label="Rightwards")
ax.plot(ioh_wk["week_from_election"], smooth(ioh_wk["prop"]),
        linewidth=2.5, color="#27ae60", label="Israel Our Home")

ax.axvline(x=0, color="#2c3e50", linestyle="--", linewidth=2,
           label="Election — March 23, 2021")

ax.yaxis.set_major_formatter(mtick.PercentFormatter(xmax=1, decimals=1))
ax.set_xlabel("Weeks from Election (March 23, 2021)", fontweight="bold")
ax.set_ylabel("Proportion of Populist Tweets", fontweight="bold")
ax.set_title("Populist Rhetoric Over Time:\nPRRPs in Change Coalition (Rightwards vs Israel Our Home)",
             fontweight="bold", pad=14)
ax.legend(loc="upper left", frameon=True, fancybox=True, shadow=True, fontsize=10)
ax.grid(True, alpha=0.3)
plt.tight_layout()
plt.savefig("output/plot2b_change_coalition_trend.png", bbox_inches="tight")
plt.close()
print("✓ output/plot2b_change_coalition_trend.png")

print("\n=== STAGE 2 COMPLETE ===")

File: 02_time_series_plots.R
# ==============================================================================
# Stage 2: Time Series Plots (R)
# ==============================================================================
library(tidyverse)
library(zoo)   # rollmean
options(scipen = 999)

cat("=== STAGE 2: TIME SERIES PLOTS ===\n\n")

data <- read_csv("output/analysis_data.csv", show_col_types = FALSE)
WINDOW <- 4

weekly_prop <- function(df) {
  df %>%
    group_by(week_from_election) %>%
    summarise(prop = mean(pop), .groups = "drop") %>%
    arrange(week_from_election)
}

election_line <- geom_vline(xintercept = 0, linetype = "dashed",
                             color = "#2c3e50", linewidth = 1.2)
pct_fmt <- scales::percent_format(accuracy = 0.1)

# ── Plot 2a ──────────────────────────────────────────────────────────────────
cat("Plot 2a: All parties vs Radicalized...\n")

all_wk <- weekly_prop(data) %>%
  mutate(smooth = rollmean(prop, k = WINDOW, fill = NA, align = "center"),
         group = "All parties")

rad_wk <- weekly_prop(data %>% filter(radicalized_group)) %>%
  mutate(smooth = rollmean(prop, k = WINDOW, fill = NA, align = "center"),
         group = "Radicalized and Radical Populism\n(Likud + Religious Zionism)")

ts_data <- bind_rows(all_wk, rad_wk)

colors2a <- c("All parties" = "#7f8c8d",
              "Radicalized and Radical Populism\n(Likud + Religious Zionism)" = "#e74c3c")

p2a <- ggplot(ts_data, aes(x = week_from_election, color = group)) +
  geom_point(aes(y = prop), alpha = 0.18, size = 1.2) +
  geom_line(aes(y = smooth), linewidth = 1.8, na.rm = TRUE) +
  election_line +
  scale_color_manual(values = colors2a, name = NULL) +
  scale_y_continuous(labels = pct_fmt) +
  labs(x = "Weeks from Election (March 23, 2021)",
       y = "Proportion of Populist Tweets",
       title = "Populist Rhetoric Over Time:",
       subtitle = "All Parties vs Radicalized and Radical Populism") +
  theme_minimal() +
  theme(plot.title    = element_text(face = "bold", size = 14),
        plot.subtitle = element_text(size = 12),
        axis.title    = element_text(face = "bold"),
        legend.position = "top",
        panel.grid.minor = element_blank())

ggsave("output/plot2a_radicalized_trend.png", p2a, width = 12, height = 6, dpi = 300)
cat("✓ output/plot2a_radicalized_trend.png\n")

# ── Plot 2b ──────────────────────────────────────────────────────────────────
cat("\nPlot 2b: Rightwards vs Israel Our Home...\n")

right_wk <- weekly_prop(data %>% filter(party == "Rightwards")) %>%
  mutate(smooth = rollmean(prop, k = WINDOW, fill = NA, align = "center"),
         group = "Rightwards")

ioh_wk <- weekly_prop(data %>% filter(party == "Israel Our Home")) %>%
  mutate(smooth = rollmean(prop, k = WINDOW, fill = NA, align = "center"),
         group = "Israel Our Home")

ts2b <- bind_rows(right_wk, ioh_wk)
colors2b <- c("Rightwards" = "#2980b9", "Israel Our Home" = "#27ae60")

p2b <- ggplot(ts2b, aes(x = week_from_election, color = group)) +
  geom_point(aes(y = prop), alpha = 0.18, size = 1.2) +
  geom_line(aes(y = smooth), linewidth = 1.8, na.rm = TRUE) +
  election_line +
  scale_color_manual(values = colors2b, name = NULL) +
  scale_y_continuous(labels = pct_fmt) +
  labs(x = "Weeks from Election (March 23, 2021)",
       y = "Proportion of Populist Tweets",
       title = "Populist Rhetoric Over Time:",
       subtitle = "PRRPs in Change Coalition (Rightwards vs Israel Our Home)") +
  theme_minimal() +
  theme(plot.title    = element_text(face = "bold", size = 14),
        plot.subtitle = element_text(size = 12),
        axis.title    = element_text(face = "bold"),
        legend.position = "top",
        panel.grid.minor = element_blank())

ggsave("output/plot2b_change_coalition_trend.png", p2b, width = 12, height = 6, dpi = 300)
cat("✓ output/plot2b_change_coalition_trend.png\n")

cat("\n=== STAGE 2 COMPLETE ===\n")
